web/controllers/registration_controller.ex

```diff
@@ -0,0 +1,42 @@
+defmodule Yggdrasil.RegistrationController do
+  use Yggdrasil.Web, :controller
+
+  import Ecto.Changeset, only: [put_change: 3]
+  import Comeonin.Bcrypt, only: [hashpwsalt: 1]
+
+  alias Yggdrasil.User
+
+  # if logged in, go straight to client
+  plug Yggdrasil.Plug.SkipIfAuthenticated
+  plug :scrub_params, "user" when action in [:create]
+
+  def new(conn, _params) do
+    changeset = User.changeset(%User{})
+    render conn, changeset: changeset
+  end
+
+  def create(conn, %{"user" => user_params}) do
+    changeset = User.changeset(%User{}, user_params)
+
+    # this whole block needs to be simplified
+    # copied from various sources
+    if changeset.valid? do
```

Comment:
So is it convention to check the changeset here and below? The Ecto [changeset docs](http://hexdocs.pm/ecto/Ecto.Changeset.html) have an example in the `Validations and constraints` section where they pass it to the repo directly and check the result. It appears to return validation errors before attempting to hit the database.